Various robustness additions to the Davis EOS

jhp-lanl opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "fix bug in ideal gas EOS.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary
Three primary additions to the Davis Reactants and Products EOS to make them more robust:
1. Added the `robust::ratio` function when dividing by density
2. Added early returns for non-positive densities. Since the products EOS should exhibit ideal gas behavior at zero density, it makes sense to return zeros for the reference values.
3. Maxed the density with zero in the reactants EOS. With the robust ratio this basically just returns a low density limit. I don't think there's an overflow issue, but I could be wrong.

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- N/A Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file.
- [ ] LANL employees: make sure tests pass both on the github CI and on the Darwin CI

If preparing for a new release, in addition please check the following:
- [ ] Update the version in cmake.
- [ ] Move the changes in the CHANGELOG.md file under a new header for the new release, and reset the categories.

jhp-lanl commented:
@chadmeyer please review these changes and see if you agree with my approach

chadmeyer commented on singularity-eos/eos/eos_davis.hpp:
```diff
@@ -328,7 +354,7 @@ class DavisProducts : public EosBase<DavisProducts> {
 PORTABLE_INLINE_FUNCTION Real DavisReactants::Ps(const Real rho) const {
   using namespace math_utils;
-  const Real y = 1.0 - _rho0 / rho;
+  const Real y = 1.0 - robust::ratio(_rho0, rho);
```
Also see comment below.  If rho is slightly negative, y will tend to $+\infty$.  Is this a problem?  If rho is zero, then the value should be `-phat`, but it will be slightly not that (I doubt it matters unless there is some kind of numerical integration scheme being used on this routine).

Yurlungur replied:
You could enforce `rho` positivity by doing `robust::ratio(_rho0, std::max(0, rho));` or something similar

chadmeyer commented on singularity-eos/eos/eos_davis.hpp:
```diff
@@ -342,7 +368,7 @@ PORTABLE_INLINE_FUNCTION Real DavisReactants::Ps(const Real rho) const {
 }
 PORTABLE_INLINE_FUNCTION Real DavisReactants::Es(const Real rho) const {
-  const Real y = 1 - _rho0 / rho;
+  const Real y = 1 - robust::ratio(_rho0, std::max(rho, 0.));
```
Is this (max(rho,0)) desired?  Is it necessary to guard against negative rho?  If so, we should be consistent between above and here.

jhp-lanl replied:
Yes that is the intent. I'll correct the above

chadmeyer commented on singularity-eos/eos/eos_davis.hpp:
```diff
@@ -354,19 +380,21 @@ PORTABLE_INLINE_FUNCTION Real DavisReactants::Es(const Real rho) const {
-    return _T0 * std::exp(-_Z * y) * std::pow(_rho0 / rho, -_G0 - _Z);
+    const Real y = 1 - robust::ratio(_rho0, std::max(rho, 0.));
+    return _T0 * std::exp(-_Z * y) *
+           std::pow(robust::ratio(_rho0, std::max(rho, 0.)), -_G0 - _Z);
```
In this branch rho is always > 0.

chadmeyer commented on singularity-eos/eos/eos_davis.hpp:
```diff
@@ -354,19 +380,21 @@ PORTABLE_INLINE_FUNCTION Real DavisReactants::Es(const Real rho) const {
+           std::pow(robust::ratio(_rho0, std::max(rho, 0.)), -_G0 - _Z);
   } else {
-    return _T0 * std::pow(_rho0 / rho, -_G0);
+    return _T0 * std::pow(robust::ratio(_rho0, rho), -_G0);
```
In this branch where rho could go to zero, it might make sense to make sure rho is nonnegative.  In the end, you might just want to store `robust::ratio(_rho0.,std::max(rho,0.0))` at the beginning of this routine since it is re-used multiple times.

jhp-lanl replied:
For this comment and the previous, I had eliminated negative density branches in favor of just forcing a positive density. I'll do that here too

chadmeyer commented on singularity-eos/eos/eos_davis.hpp:
```diff
@@ -389,15 +417,18 @@ PORTABLE_INLINE_FUNCTION Real DavisReactants::BulkModulusFromDensityInternalEner
   const Real gammav = (rho >= _rho0) ? _Z * _rho0 : 0.0;
-  return -(psv + (sie - Es(rho)) * rho * (gammav - gamma * rho) - gamma * rho * esv) /
-         rho;
+  const Real numerator =
+      -(psv + (sie - Es(rho)) * rho * (gammav - gamma * std::max(rho, 0.)) -
+        gamma * std::max(rho, 0.) * esv);
+  return robust::ratio(numerator, rho);
```
Is there a reason for some of the rhos having a max and some not?

jhp-lanl replied:
No... I was just probably working too fast. I've added `max` to all `rho` now here

jhp-lanl commented on singularity-eos/eos/eos_davis.hpp:
```diff
@@ -72,8 +79,11 @@ class DavisReactants : public EosBase<DavisReactants> {
+    // when the base to the exponent is zero (see T(rho, e) equation)
+    const Real es = Es(rho);
+    const Real ts = Ts(rho);
+    return es - (1 + _alpha) / (_Cv0 * ts);
```
Ah good point. Thanks for the catch. I flipped that in my math. I'll change shortly

Yurlungur commented on singularity-eos/eos/eos_davis.hpp:
```diff
@@ -100,8 +110,12 @@ class DavisReactants : public EosBase<DavisReactants> {
       Indexer_t &&lambda = static_cast<Real *>(nullptr)) const {
-    return _Cv0 / std::pow((1 + _alpha) / (Ts(rho) * _Cv0) * (sie - Es(rho)) + 1,
-                           -_alpha / (1 + _alpha));
+    const Real power_base = (1 + _alpha) / (Ts(rho) * _Cv0) * (sie - Es(rho)) + 1;
```
Since this base is now being explicitly pulled out and checked there's maybe some opportunity to unify it with the other places it's called.

jhp-lanl replied:
I broke it out into a function

Yurlungur commented on singularity-eos/eos/eos_davis.hpp:
```diff
@@ -418,6 +449,7 @@ PORTABLE_INLINE_FUNCTION void DavisReactants::FillEos(Real &rho, Real &temp, Rea
                                                       Real &press, Real &cv, Real &bmod,
                                                       const unsigned long output,
                                                       Indexer_t &&lambda) const {
+  // BROKEN: This can only handle density-energy inputs! MUST BE CHANGED
```
Good catch. Yes we should fix that.

Yurlungur reviewed (approved):
I defer to @chadmeyer 's comments. But other than what chad picked out I'm happy. Thanks for the fixes.

jhp-lanl commented:
@chadmeyer I think I addressed all your comments. I'll give things another scan and make sure I didn't miss anything else silly, but I think this is good now.

jhp-lanl commented:
FYI - I added a few more robustness additions. Mainly I added an error to the `DensityEnergyFromPressureTemperature()` functions if the resultant density is negative (adding a max here probably will just cause the function to not converge and produce a less informative error).

I also added a `robust::ratio` to where we divide by `Ts`. It is possible for this to underflow which _might_ be zero.
